### PR TITLE
fetchcargo: use flat tar.gz file for vendored src instead of recursive hash dir

### PR DESCRIPTION
**Pull request by bhipple**
This has several advantages:

1. It takes up significantly less space on disk in-between builds in the nix store.
2. It plays nicely with hashed mirrors like tarballs.nixos.org, which only
   substitute --flat hashes on single files (not recursive directory hashes).
3. It's consistent with how simple `fetchurl` src derivations work.
4. It provides a stronger abstraction between input src-package and output
   package, e.g., it's harder to accidentally depend on the src derivation at
   runtime by referencing something like `${src}/etc/index.html`. Likewise, in
   the store it's harder to get confused with something that is just there as a
   build-time dependency vs. a runtime dependency, since the build-time
   src dependencies are tarred up.

Note that since the binary cache vendor packages are already compressed with xz, it's not more efficient on that end or in terms of network traffic.

Disadvantages are:
1. It takes slightly longer to untar at the start of a build (~2 seconds).

As currently implemented, this attaches the compacted vendor.tar.gz feature as a
rider on `verifyCargoDeps`, since both of them are relatively newly implemented
behavior that change the `cargoSha256`.

If this PR is accepted, I will push forward the remaining rust packages with a
series of treewide PRs to update the `cargoSha256`s.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bhipple** [comment]
@GrahamcOfBorg build broot ripgrep

**Comment by bhipple** [comment]
CC @grahamc @zimbatm

File: pkgs/build-support/rust/README.md
Comment (bhipple) [inline pkgs/build-support/rust/README.md]: I have updated the hash on every package that has `verifyCargoDeps = true;` in this PR, and verified that other rust packages with this set to false are not impacted, e.g., changing 1 character on the hash for `ripgrep` and re-building it produces an error that shows the old hash.  

I have a script to do the rest of them and set it to true everywhere, but figured I'd open the PR first for feedback about whether others agree this is a good idea in general first.  If so, once it's merged I can send the treewide follow-up PR that changes all the hashes and sets the attribute to true, then once that's merged I'll send another that removes the boolean.

**Review by zimbatm — COMMENTED** [review]
I think it's a good idea.

Since this is a breaking change to the output hash, can you rename the `verifyCargoDeps` variable to something else like `archiveCargoDeps`? That way external users will get an explicit break rather than mysteriously changing hashes.

Since fetchcargo now produces two entirely different outputs, what do you think of creating a new `fetchCargoTarball` fetcher instead? There would be a bit of duplication but it would also avoid having that logic inter-linked.

**Comment by bhipple** [comment]
Nice, I'll play around with the idea of creating an entirely different `fetchcargo` helper and leaving the old one modified and see how it looks.  Either way I intend to get us back to just having 1 standard way of doing things and hope to delete the legacy behavior once ported, but tactically your suggestion might be superior.

**Comment by bhipple** [comment]
Ok, per discussion I have refactored to leave the old `fetchcargo` alone, and implement this as a switch on a new, separate `fetchCargoTarball`.  This does look cleaner, and it makes it clearer what the new implementation looks like once all the legacy stuff is deleted.

File: pkgs/build-support/rust/fetchCargoTarball.nix
Comment (bhipple) [inline pkgs/build-support/rust/fetchCargoTarball.nix]: I've run this through a bunch of times and verified that it is indeed reproducible.

File: pkgs/build-support/rust/default.nix
Comment (bhipple) [inline pkgs/build-support/rust/default.nix]: While we're making some light changes in the default.nix, there is no change in `fetchcargo.nix` -- just a switch on which implementation to use.

File: pkgs/build-support/rust/README.md
Comment (bhipple) [inline pkgs/build-support/rust/README.md]: I think the actual implementation is relatively straightforward and the advantages are clear, but I could foresee this causing git merge conflict hell for maintainers.  As implemented here and discussed I think this provides the easiest/clearest way for maintainers to resolve cherry-pick conflicts on stable backports, but if anyone has any better guidance LMK.

**Review by bhipple — COMMENTED** [review]
Building one thing with the new behavior and one thing without:

@GrahamcOfBorg build ripgrep hexdino

As a tip to anyone testing, `hexdino` appears to be one of the smallest packages with the fewest vendored dependencies to download and compile, so it's great for fast edit -> compile -> test cycles.

File: pkgs/build-support/rust/fetchCargoTarball.nix
Comment (misuzu) [inline pkgs/build-support/rust/fetchCargoTarball.nix]: I'd remove `v` flag.

Comment (bhipple) [reply pkgs/build-support/rust/fetchCargoTarball.nix]: Removed.  BTW, here's a concrete example of the difference that this makes to the `attr.cargoDeps` size:

```
# Ripgrep's cargoDeps as currently built on master:
156M    /nix/store/vdaa161ixsyjmvqpyf101nvkrf9w0ddm-ripgrep-11.0.2-vendor

# Ripgrep's cargoDeps if we were to set legacyCargoFetcher = false on this PR and update hash:
16M     /nix/store/cwafl8f9gradpyhs4mjygnkl8mijq558-ripgrep-11.0.2-vendor.tar.gz
```
156M -> 16M is a pretty big difference when you add up across all the Rust packages!

Comment (zimbatm) [reply pkgs/build-support/rust/fetchCargoTarball.nix]: I don't think it's worth spending too much CPU on this. The NAR files are already compressed (with xz) and the end-user installing ripgrep won't be downloading this.

Comment (bhipple) [reply pkgs/build-support/rust/fetchCargoTarball.nix]: Sounds good, I'll leave it as is then.  

This does cause a rebuild of all the Rust applications, due to the attributes shuffling around, but no `cargoSha256`s or versions are materially changed, so I've currently got it going against `master`, but I can rebase and retarget for `staging` if we prefer.

Comment (zimbatm) [reply pkgs/build-support/rust/fetchCargoTarball.nix]: Yeah I think this is due for staging due to the amount of rebuild. It might also make sense to do the whole migration as a single commit to reduce the amount of rebuild.

File: pkgs/build-support/rust/default.nix
Comment (zimbatm) [inline pkgs/build-support/rust/default.nix]:
```suggestion
  '' + stdenv.lib.optionalString (cargoDeps.passthru ? checker) cargoDeps.passthru.checker 
```
This is just an idea I want to mention, it doesn't need to be acted upon.

If that code was moved to the fetcher, it would allow to pass `cargoDeps` to the buildRustPackage and make this whole thing more composable. Right now the arguments to buildRustPackage contain both arguments for it and the fetcher which makes it quite wide. By separating the cargoDeps it would allow for example to change the compression of the deps in a local development scenario.

It's just an idea. It also has downsides since some of the arguments like the `src` are currently shared.

**Review by zimbatm — APPROVED** [review]

File: pkgs/build-support/rust/fetchCargoTarball.nix
Comment (bhipple) [reply pkgs/build-support/rust/fetchCargoTarball.nix]: I plan to move relatively quickly on this once it lands on master, but ideally would like to proceed in the incremental steps outlined in the doc to attempt to reduce merge conflicts.

Already I had to carry 1 commit over from master to resolve a merge conflict for the git-workspace tool :)

File: pkgs/build-support/rust/default.nix
Comment (bhipple) [reply pkgs/build-support/rust/default.nix]: Hmm, I'll ponder it over -- but in the meantime, while looking at this area of the code and while prepping some work for the next step in the mass migration where we set `legacyCargoFetcher = false;` by default, I noticed that this block of code gets entered into even when `cargoSha256 == "unset";`, e.g. building `cargo` itself.  I've updated accordingly to check that we actually have a `cargoSha256` before validating it now.

File: pkgs/build-support/rust/default.nix
Comment (bhipple) [inline pkgs/build-support/rust/default.nix]: Also added this, which ensures that the default value of `legacyCargoFetcher` does not impact the hash sent to `stdenv.mkDerivation`.  This lets us do the next step (default it to true, opting out all packages via a treewide sed) without causing any hash changes or package rebuilds, including in packages like `cargo` that do use the Rust build platform infra but don't use any fetchers (it inherits its vendor dir from `rustc`).

File: pkgs/build-support/rust/README.md
Comment (bhipple) [reply pkgs/build-support/rust/README.md]: A script that accomplishes this is attached at the bottom.  I have run it on this implementation and verified that it does not cause any rebuilds or change any hashes.  Since it touches 200+ files and causes no rebuilds, my current plan is to let the implementation filter its way through staging -> staging-next -> master, then send the treewide PR to master.  

At that point, all of the Rust applications will be explicitly opting into the legacy Cargo fetcher, with a comment to maintainers to delete the attr line on the next update.  

```bash
#!/usr/bin/env zsh
set -euo pipefail
cd ~/src/nixpkgs || exit 1

sed -i 's|^, legacyCargoFetcher.*|, legacyCargoFetcher ? false|' pkgs/build-support/rust/default.nix

for f in $(rg -l 'cargoSha256 = "' **/*.nix); do
    if rg -q 'legacyCargoFetcher = false;' $f; then
        continue
    fi

    sed -i '/cargoSha256 = "/i   # Please delete this line on next update\n  legacyCargoFetcher = true;\n' $f
done
```

**Comment by bhipple** [comment]
Rebased onto origin/staging to fix a merge conflict, but otherwise unchanged.

**Comment by bhipple** [comment]
Thanks for the feedback and review @zimbatm.  Anyone else mind taking a look and giving a review?  @andir perhaps?

**Comment by zimbatm** [comment]
@GrahamcOfBorg eval

**Comment by zimbatm** [comment]
@bhipple do you see any ways to change the PR so that it's not a mass rebuild? If not then let's merge this to staging.

**Comment by bhipple** [comment]
Unfortunately I don't -- though this does clean the fetcher impl used from the stdenv.mkDerivation, so the follow-ups will not be mass rebuilds.

**Comment by bhipple** [comment]
I rebased onto staging and cherry-picked in the broot update so I could re-compute the sha, but made no other material changes.  I think this will (probably?) still cause git merge conflicts on broot, but as stacked my final commit has the "right" value.

**Comment by zimbatm** [comment]
should we merge this?

**Comment by bhipple** [comment]
Rebased again to fix `broot` merge conflict.  Yes I'd say let's merge this :)

**Comment by zimbatm** [comment]
please make sure to monitor the staging channel to see if there are any regressions :)

**Comment by petabyteboy** [comment]
So while updating my systems I ran into a hash mismatch for a third party derivation for [shelfie](https://github.com/petabyteboy/shelfie/blob/master/default.nix).

The errors were very confusing to me. First I got this:

```
hash mismatch in fixed-output derivation '/nix/store/6kn0lgbb42pa51i2p41aqhdpb3klz07n-shelfie-0.1.0-vendor.tar.gz':
  wanted: sha256:06nhwxyv2x78gdky5lz9raipx6z764n9ffqwq30x4mfxbsbdwf2x
  got:    sha256:0085z26g6j80nf0zbwqxnbcqyjhxzknvb5wy21fmrrmy4ldrp4fl
```

So I changed the hash and rebuilt the system, which gave me this error:

```
builder for '/nix/store/kaxfdgqak5y8hmc3q314x53cs8hgbwy4-shelfie-0.1.0.drv' failed with exit code 1; last 10 log lines:

  ERROR: cargoSha256 is out of date

  Cargo.lock is not the same in shelfie-0.1.0-vendor.tar.gz

  To fix the issue:
  1. Use "1111111111111111111111111111111111111111111111111111" as the cargoSha256 value
  2. Build the derivation and wait it to fail with a hash mismatch
  3. Copy the 'got: sha256:' value back into the cargoSha256 field
```

But that's literally what I just did.
What is going on here?

**Comment by bhipple** [comment]
See here: https://github.com/NixOS/nixpkgs/pull/80262/files/f180adc0c1cdba75a5fd540d8a86880e36aade0b#r379944118

Because you're using `src = ./.;`, the name of the source package is not "source", so it's failing to find the Cargo.lock to compare against the cargoSha256 vendor dir.  Can you try rebuilding shelfie with that PR and see if it "just works"?

That's on the staging branch, so cheaper/faster alternatives are to fix it with either the LHS or RHS of a diff like this:
https://github.com/NixOS/nixpkgs/pull/80239/files

E.g., you can either use the legacy cargo fetcher, which won't validate that your src and cargoSha256 are in sync, or you can `ln` the source dir to be explicitly named `source`.

**Comment by petabyteboy** [comment]
I see! It works fine after cherry-picking the commit from the PR you mentioned, thanks.

**Comment by zimbatm** [comment]
one way to force the source name is to use `builtins.path` like this: `src = builtins.path { name = "source"; path = ./.; }`

**Comment by bhipple** [comment]
The root "issue" here is that `fetchFromGitHub` (and a couple other fetchers) will give the fixed-output derivation the name `source` if no other is provided, which almost none of them do:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgithub/default.nix#L3

However, since users *can* provide a different name, and since a couple of the fetchers do, I think it's good practice for us to not rely on that particular implementation detail of `fetchFromGitHub`.  We do have workarounds like `ln -s ${name} source`, which we can do at an individual package level, but I don't think we can do that at an infra level, since we don't know (nor should we) the given name of the `src` attribute's package.  

Using a glob to be generic seems like the right approach to me philosophically; the trick is thinking of what weird edge conditions might exist and making it robust.

**Pull request merged**
